File: auth/go/aws/callback.go
package aws

import (
	"encoding/json"
	"io"
	"net/http"

	"authlab/shared"
)

func Callback(w http.ResponseWriter, r *http.Request) {
	sid := shared.Sessions.GetOrCreate(w, r)
	expectedState, _ := shared.Sessions.Get(sid, "state_aws")
	if r.URL.Query().Get("state") != expectedState {
		http.Error(w, "state invalido", http.StatusBadRequest)
		return
	}

	code := r.URL.Query().Get("code")
	token, err := cfg.Exchange(r.Context(), code)
	if err != nil {
		http.Error(w, err.Error(), http.StatusInternalServerError)
		return
	}

	// Igual que en las otras implementaciones: se pide el userinfo
	// endpoint con el access_token en vez de decodificar el id_token.
	req, _ := http.NewRequestWithContext(r.Context(), http.MethodGet, userInfoURL, nil)
	req.Header.Set("Authorization", "Bearer "+token.AccessToken)
	resp, err := http.DefaultClient.Do(req)
	if err != nil {
		http.Error(w, err.Error(), http.StatusInternalServerError)
		return
	}
	defer resp.Body.Close()
	raw, _ := io.ReadAll(resp.Body)

	var claims map[string]any
	if err := json.Unmarshal(raw, &claims); err != nil {
		http.Error(w, "respuesta userinfo invalida: "+string(raw), http.StatusInternalServerError)
		return
	}

	shared.Sessions.Set(sid, "user_aws", claims)
	http.Redirect(w, r, "/aws/private", http.StatusFound)
}

File: auth/php/aws/callback.php
<?php

function aws_callback(): void
{
    $config = aws_config();

    if (($_GET["state"] ?? null) !== ($_SESSION["state_aws"] ?? null)) {
        http_response_code(400);
        echo layout("<pre>state invalido</pre>");
        exit;
    }

    $token = httpPostForm($config["token_endpoint"], [
        "grant_type" => "authorization_code",
        "code" => $_GET["code"] ?? "",
        "redirect_uri" => $config["redirect_uri"],
        "client_id" => $config["client_id"],
        "client_secret" => $config["client_secret"],
    ]);

    if (empty($token["access_token"])) {
        http_response_code(500);
        echo layout("<pre>" . htmlspecialchars(json_encode($token)) . "</pre>");
        exit;
    }

    // Igual que en las otras implementaciones: se pide el userinfo
    // endpoint con el access_token en vez de decodificar el id_token.
    $_SESSION["user_aws"] = httpGetBearer($config["userinfo_endpoint"], $token["access_token"]);
    redirectTo("/aws/private");
}
